`TestShutdownDeadlock` sporadically fails

```sh
git log -n 1 --oneline
d6f325f (HEAD -> master, origin/master, origin/HEAD) Include messages in distributed results (#115)

go test -v -run TestShutdownDeadlock 2>&1 | tee out.txt

=== RUN   TestShutdownDeadlock
=== RUN   TestShutdownDeadlock/#00
=== PAUSE TestShutdownDeadlock/#00
=== RUN   TestShutdownDeadlock/#01
=== PAUSE TestShutdownDeadlock/#01
=== RUN   TestShutdownDeadlock/#02
=== PAUSE TestShutdownDeadlock/#02
=== RUN   TestShutdownDeadlock/#03
=== PAUSE TestShutdownDeadlock/#03
=== RUN   TestShutdownDeadlock/#04
=== PAUSE TestShutdownDeadlock/#04
=== RUN   TestShutdownDeadlock/#05
=== PAUSE TestShutdownDeadlock/#05
=== RUN   TestShutdownDeadlock/#06
=== PAUSE TestShutdownDeadlock/#06
=== RUN   TestShutdownDeadlock/#07
=== PAUSE TestShutdownDeadlock/#07
=== RUN   TestShutdownDeadlock/#08
=== PAUSE TestShutdownDeadlock/#08
=== RUN   TestShutdownDeadlock/#09
=== PAUSE TestShutdownDeadlock/#09
=== RUN   TestShutdownDeadlock/#10
=== PAUSE TestShutdownDeadlock/#10
=== RUN   TestShutdownDeadlock/#11
=== PAUSE TestShutdownDeadlock/#11
=== RUN   TestShutdownDeadlock/#12
=== PAUSE TestShutdownDeadlock/#12
=== RUN   TestShutdownDeadlock/#13
=== PAUSE TestShutdownDeadlock/#13
=== RUN   TestShutdownDeadlock/#14
=== PAUSE TestShutdownDeadlock/#14
=== RUN   TestShutdownDeadlock/#15
=== PAUSE TestShutdownDeadlock/#15
=== RUN   TestShutdownDeadlock/#16
=== PAUSE TestShutdownDeadlock/#16
=== RUN   TestShutdownDeadlock/#17
=== PAUSE TestShutdownDeadlock/#17
=== RUN   TestShutdownDeadlock/#18
=== PAUSE TestShutdownDeadlock/#18
=== RUN   TestShutdownDeadlock/#19
=== PAUSE TestShutdownDeadlock/#19
=== CONT  TestShutdownDeadlock/#00
=== CONT  TestShutdownDeadlock/#10
=== CONT  TestShutdownDeadlock/#15
=== CONT  TestShutdownDeadlock/#13
=== CONT  TestShutdownDeadlock/#05
=== CONT  TestShutdownDeadlock/#08
=== CONT  TestShutdownDeadlock/#14
=== CONT  TestShutdownDeadlock/#03
=== CONT  TestShutdownDeadlock/#02
=== CONT  TestShutdownDeadlock/#01
=== CONT  TestShutdownDeadlock/#04
=== CONT  TestShutdownDeadlock/#18
=== CONT  TestShutdownDeadlock/#19
=== CONT  TestShutdownDeadlock/#09
=== CONT  TestShutdownDeadlock/#17
=== CONT  TestShutdownDeadlock/#07
=== CONT  TestShutdownDeadlock/#12
=== CONT  TestShutdownDeadlock/#16
=== CONT  TestShutdownDeadlock/#11
=== CONT  TestShutdownDeadlock/#06
=== NAME  TestShutdownDeadlock/#03
    server_test.go:175: hung on shutdown
=== NAME  TestShutdownDeadlock/#09
    server_test.go:175: hung on shutdown
--- FAIL: TestShutdownDeadlock (0.00s)
    --- PASS: TestShutdownDeadlock/#00 (0.02s)
    --- PASS: TestShutdownDeadlock/#15 (0.02s)
    --- PASS: TestShutdownDeadlock/#05 (0.02s)
    --- PASS: TestShutdownDeadlock/#14 (0.02s)
    --- PASS: TestShutdownDeadlock/#08 (0.02s)
    --- PASS: TestShutdownDeadlock/#10 (0.02s)
    --- PASS: TestShutdownDeadlock/#02 (0.01s)
    --- PASS: TestShutdownDeadlock/#04 (0.01s)
    --- PASS: TestShutdownDeadlock/#01 (0.01s)
    --- PASS: TestShutdownDeadlock/#18 (0.01s)
    --- PASS: TestShutdownDeadlock/#12 (0.01s)
    --- PASS: TestShutdownDeadlock/#07 (0.01s)
    --- PASS: TestShutdownDeadlock/#17 (0.01s)
    --- PASS: TestShutdownDeadlock/#06 (0.01s)
    --- PASS: TestShutdownDeadlock/#11 (0.01s)
    --- PASS: TestShutdownDeadlock/#13 (0.52s)
    --- PASS: TestShutdownDeadlock/#19 (0.51s)
    --- PASS: TestShutdownDeadlock/#16 (0.51s)
    --- FAIL: TestShutdownDeadlock/#03 (5.02s)
    --- FAIL: TestShutdownDeadlock/#09 (5.01s)
FAIL
exit status 1
FAIL	github.com/osquery/osquery-go	5.282s

```

--- a/server.go
+++ b/server.go
@@ -322,6 +322,7 @@ func (s *ExtensionManagerServer) Call(ctx context.Context, registry string, item
 func (s *ExtensionManagerServer) Shutdown(ctx context.Context) (err error) {
 	s.mutex.Lock()
 	defer s.mutex.Unlock()
+
 	stat, err := s.serverClient.DeregisterExtension(s.uuid)
 	err = errors.Wrap(err, "deregistering extension")
 	if err == nil && stat.Code != 0 {
@@ -333,7 +334,7 @@ func (s *ExtensionManagerServer) Shutdown(ctx context.Context) (err error) {
 		s.server = nil
 		// Stop the server asynchronously so that the current request
 		// can complete. Otherwise, this is vulnerable to deadlock if a
-		// shutdown request is being processed when shutdown is
+		// shutdown request is being processed when Shutdown is
 		// explicitly called.
 		go func() {
 			server.Stop()

--- a/server_test.go
+++ b/server_test.go
@@ -7,6 +7,7 @@ import (
 	"io/ioutil"
 	"net"
 	"os"
+	"runtime/pprof"
 	"strings"
 	"sync"
 	"syscall"
@@ -98,18 +99,19 @@ const parallelTestShutdownDeadlock = 20
 
 func TestShutdownDeadlock(t *testing.T) {
 	for i := 0; i < parallelTestShutdownDeadlock; i++ {
+		i := i
 		t.Run("", func(t *testing.T) {
 			t.Parallel()
-			testShutdownDeadlock(t)
+			testShutdownDeadlock(t, i)
 		})
 	}
 }
-func testShutdownDeadlock(t *testing.T) {
+func testShutdownDeadlock(t *testing.T, uuid int) {
 	tempPath, err := ioutil.TempFile("", "")
 	require.Nil(t, err)
 	defer os.Remove(tempPath.Name())
 
-	retUUID := osquery.ExtensionRouteUUID(0)
+	retUUID := osquery.ExtensionRouteUUID(uuid)
 	mock := &MockExtensionManager{
 		RegisterExtensionFunc: func(info *osquery.InternalExtensionInfo, registry osquery.ExtensionRegistry) (*osquery.ExtensionStatus, error) {
 			return &osquery.ExtensionStatus{Code: 0, UUID: retUUID}, nil
@@ -119,16 +121,22 @@ func testShutdownDeadlock(t *testing.T) {
 		},
 		CloseFunc: func() {},
 	}
-	server := ExtensionManagerServer{serverClient: mock, sockPath: tempPath.Name()}
+	server := ExtensionManagerServer{
+		serverClient: mock,
+		sockPath:     tempPath.Name(),
+		timeout:      defaultTimeout,
+	}
 
-	wait := sync.WaitGroup{}
+	var wait sync.WaitGroup
 
-	wait.Add(1)
 	go func() {
+		// We do not wait for this routine to finish because thrift.TServer.Serve
+		// seems to sometimes hang after shutdowns. (This test is just testing
+		// the Shutdown doesn't hang.)
 		err := server.Start()
-		require.Nil(t, err)
-		wait.Done()
+		require.NoError(t, err)
 	}()
+
 	// Wait for server to be set up
 	server.waitStarted()
 
@@ -138,10 +146,17 @@ func testShutdownDeadlock(t *testing.T) {
 	addr, err := net.ResolveUnixAddr("unix", listenPath)
 	require.Nil(t, err)
 	timeout := 500 * time.Millisecond
-	trans := thrift.NewTSocketFromAddrTimeout(addr, timeout, timeout)
-	err = trans.Open()
-	require.Nil(t, err)
-	client := osquery.NewExtensionManagerClientFactory(trans,
+	opened := false
+	attempt := 0
+	var transport *thrift.TSocket
+	for !opened && attempt < 10 {
+		transport = thrift.NewTSocketFromAddrTimeout(addr, timeout, timeout)
+		err = transport.Open()
+		opened = err == nil
+		attempt++
+	}
+	require.NoError(t, err)
+	client := osquery.NewExtensionManagerClientFactory(transport,
 		thrift.NewTBinaryProtocolFactoryDefault())
 
 	// Simultaneously call shutdown through a request from the client and
@@ -156,7 +171,7 @@ func testShutdownDeadlock(t *testing.T) {
 	go func() {
 		defer wait.Done()
 		err = server.Shutdown(context.Background())
-		require.Nil(t, err)
+		require.NoError(t, err)
 	}()
 
 	// Track whether shutdown completed
@@ -171,7 +186,8 @@ func testShutdownDeadlock(t *testing.T) {
 	select {
 	case <-completed:
 		// Success. Do nothing.
-	case <-time.After(5 * time.Second):
+	case <-time.After(10 * time.Second):
+		pprof.Lookup("goroutine").WriteTo(os.Stdout, 1)
 		t.Fatal("hung on shutdown")
 	}
 }

